poly1305/src/avx2.rs

```diff
@@ -0,0 +1,1305 @@
+
+    state.r1 = r1;
+    state.r2 = r2;
+    state.r15 = r15;
+    state.r25 = r25;
+    state.p_init = 0;
+    state.leftover = 0;
+}
+
+#[inline]
+#[target_feature(enable = "avx2")]
+pub unsafe fn process(state: &mut Poly1305, input: &[u8]) {
+    let mut in_len = input.len();
+    let mut ip: *const u32 = input.as_ptr() as *const u32;
+    let mut p: vec256x3;
+
+    let r1 = state.r1;
+    let r15 = state.r15;
+    let r2 = state.r2;
+    let r25 = state.r25;
+    let r3 = red5x64(mul130(r2, r1, r15));
+    let r4 = red5x64(mul130(r2, r2, r25));
+
+    state.m.v0 = _mm256_blend_epi32(
+        r3,
+        _mm256_permutevar8x32_epi32(r2, _mm256_set_epi32(4, 3, 1, 0, 0, 0, 0, 0)),
+        0xE0,
+    );
+
+    state.m.v1 = _mm256_blend_epi32(
+        r4,
+        _mm256_permutevar8x32_epi32(r2, _mm256_set_epi32(4, 2, 0, 0, 0, 0, 0, 0)),
+        0xE0,
+    );
+
+    let r45 = _mm256_permutevar8x32_epi32(
+        _mm256_add_epi32(r4, _mm256_slli_epi32(r4, 2)),
+        _mm256_set_epi32(4, 3, 2, 1, 1, 1, 1, 1),
+    );
+    state.r4 = _mm256_blend_epi32(r4, r45, 0xE0);
+    state.r45 = _mm256_permute2x128_si256(r45, r45, 0);
+    p = align4x128(load4x128(ip));
+    in_len -= 64;
+    ip = ip.add(16);
+
+    state.p_init = 1;
+
+    if in_len >= 768 {
+        let mut r8 = red5x64(mul130(state.r4, state.r4, state.r45));
+        let mut r12 = red5x64(mul130(r8, state.r4, state.r45));
+        let mut r85 = _mm256_permutevar8x32_epi32(
+            (_mm256_add_epi32((r8), (_mm256_slli_epi32((r8), (2))))),
+            (_mm256_set_epi32(4, 3, 2, 1, 1, 1, 1, 1)),
+        );
+        let mut r125 = _mm256_permutevar8x32_epi32(
+            (_mm256_add_epi32((r12), (_mm256_slli_epi32((r12), (2))))),
+            (_mm256_set_epi32(4, 3, 2, 1, 1, 1, 1, 1)),
+        );
+        r8 = _mm256_blend_epi32((r8), (r85), (0xE0));
+        r12 = _mm256_blend_epi32((r12), (r125), (0xE0));
+        r85 = _mm256_permute2x128_si256((r85), (r85), (0));
+        r125 = _mm256_permute2x128_si256((r125), (r125), (0));
+
+        loop {
+            p = add4x130(
+                red4x130(muladd4x130P(
+                    muladd4x130P(mul4x130P(p, r12, r125), align4x128(load4x128(ip)), r8, r85),
+                    align4x128(load4x128(ip.add(16))),
+                    state.r4,
+                    state.r45,
+                )),
+                align4x128(load4x128(ip.add(32))),
+            );
+            in_len -= 192;
+            ip = ip.add(48);
+            if in_len < 192 {
+                break;
+            }
+        }
+
+        if in_len >= 128 {
+            p = add4x130(
+                red4x130(muladd4x130P(
+                    mul4x130P(p, r8, r85),
+                    align4x128(load4x128(ip)),
+                    state.r4,
+                    state.r45,
+                )),
+                align4x128(load4x128(ip.add(16))),
+            );
+            in_len -= 128;
+            ip = ip.add(32);
```

Comment:
It looks like an optimization which only wants to implement the current `UniversalHash` trait (which precludes optimizations for outputs larger than the UHF itself) could omit this entire codeblock and also any functions which are only invoked for this "wide" optimization.